include/wlr/types/wlr_data_device.h

```diff
@@ -23,33 +23,39 @@ struct wlr_data_offer {
 	struct wl_resource *resource;
 	struct wlr_data_source *source;
 
-	uint32_t dnd_actions;
-	enum wl_data_device_manager_dnd_action preferred_dnd_action;
+	uint32_t actions;
+	enum wl_data_device_manager_dnd_action preferred_action;
 	bool in_ask;
 
 	struct wl_listener source_destroy;
 };
 
 struct wlr_data_source {
-	struct wl_resource *resource;
-	struct wlr_data_offer *offer;
-	struct wlr_seat_client *seat_client;
-
+	// source metadata
 	struct wl_array mime_types;
+	int32_t actions;
+
+	// source implementation
+	void (*send)(struct wlr_data_source *source, const char *mime_type,
+		int32_t fd);
+	void (*accept)(struct wlr_data_source *source, uint32_t serial,
+		const char *mime_type);
+	void (*cancel)(struct wlr_data_source *source);
```

Comment:
Isn't this better done with the standard impl pattern?

Reply:
Yeah, I was hesitating switching to that. Putting mixed metadata and callbacks is still bothering me because the metadata fields need to be initialized and free'd so we'll end up with `wlr_data_source_impl_init` and `wlr_data_source_impl_finish`...

Reply:
We already have things of that sort going on, nbd

Reply:
Hmm, this is more tricky than I expected. If we split that into two structs, then the impl side won't be able to inspect the state side:

```c
struct wlr_data_source_state {
	bool accepted;
	struct wlr_data_offer *offer;
	struct wlr_seat_client *seat_client;

	enum wl_data_device_manager_dnd_action current_dnd_action;
	uint32_t compositor_action;
};

struct wlr_data_source {
	// source metadata
	struct wl_array mime_types;
	int32_t actions;

	// source implementation
	void (*send)(struct wlr_data_source *source, const char *mime_type,
		int32_t fd);
	void (*accept)(struct wlr_data_source *source, uint32_t serial,
		const char *mime_type);
	void (*cancel)(struct wlr_data_source *source);

	// drag'n'drop implementation
	void (*dnd_drop)(struct wlr_data_source *source);
	void (*dnd_finish)(struct wlr_data_source *source);
	void (*dnd_action)(struct wlr_data_source *source,
		enum wl_data_device_manager_dnd_action action);

	struct {
		struct wl_signal destroy;
	} events;
};
```

But I think we still want the impl to be able to inspect it (to implement security features depending on the client requesting the data for instance, or inspecting/changing the drag'n'drop action). So we could:

* Call all callbacks with the state, kinda breaks the impl pattern
* Keep the state in `wlr_data_source`, but in a child struct

Reply:
This feels a bit weird compared to usual? We normally have the state split the other way around, e.g. state is in wlr_data_source and there's another struct with implem in a pointer there e.g. something like that?
```
struct wlr_data_source;

struct wlr_data_source_impl {
	void (*send)(struct wlr_data_source *source, const char *mime_type,
		int32_t fd);
	void (*accept)(struct wlr_data_source *source, uint32_t serial,
		const char *mime_type);
	void (*cancel)(struct wlr_data_source *source);

	// drag'n'drop implementation
	void (*dnd_drop)(struct wlr_data_source *source);
	void (*dnd_finish)(struct wlr_data_source *source);
	void (*dnd_action)(struct wlr_data_source *source,
		enum wl_data_device_manager_dnd_action action);
};

struct wlr_data_source {
	// source metadata
	struct wl_array mime_types;
	int32_t actions;

	// source state
	bool accepted;
	struct wlr_data_offer *offer;
	struct wlr_seat_client *seat_client;

	enum wl_data_device_manager_dnd_action current_dnd_action;
	uint32_t compositor_action;

	struct wlr_data_source_impl *impl;
	struct {
		struct wl_signal destroy;
	} events;
};
```

(I might just as well be wrong :p)

Reply:
Yeah we generally do that, but we can't here because `mime_types` and `actions` must be populated by `wlr_seat_set_selection`'s caller.

Reply:
I'm not sure what just putting the functions in an impl vector changes for that, can't we keep doing what we do right now? The user is then responsible of setting both `impl` and the metadata (`mime_types` and `actions`); and also its own responsibility to listen to the destroy event signal if they need to free mime_types. (actually since we have an impl vector it might be more fitting to add a destroy handler function instead of the destroy signal even if that's less generic, unsure if we need multiple listeners right now)
If you want to be more generic, you can replace these two fields by an anonymous `void *data` and fit that in a substruct, impl will be free to use that as they see fit? But that'd require more changes as we're currently aware of the fields in wlr_data_device itself. I think I'm missing the point though...